File: tests/test_rings/test_d.py
# Copyright 2024-2025 Olivier Romain, Francis Blais, Vincent Girouard, Marius Trudeau
#
#    Licensed under the Apache License, Version 2.0 (the "License");
#    you may not use this file except in compliance with the License.
#    You may obtain a copy of the License at
#
#        http://www.apache.org/licenses/LICENSE-2.0
#
#    Unless required by applicable law or agreed to in writing, software
#    distributed under the License is distributed on an "AS IS" BASIS,
#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#    See the License for the specific language governing permissions and
#    limitations under the License.

"""Test the D class."""

import math

import mpmath as mp
import pytest

from qdecomp.rings import D

# Set a high precision for mpmath
mp.mp.dps = 75


@pytest.mark.parametrize("num", [0, 1, 5, 100, 73, -5, -1])
@pytest.mark.parametrize("denom", [0, 1, 5, 10, 19, 50])
def test_float(num: int, denom: int) -> None:
    """Test the float value of the D class."""
    assert math.isclose(num / 2**denom, float(D(num, denom)))

@pytest.mark.parametrize("num", [0, 1, 5, 100, 73, -5, -1])
@pytest.mark.parametrize("denom", [0, 1, 5, 10, 19, 50])
def test_mpfloat(num: int, denom: int) -> None:
    """Test the mpfloat value of the D class."""
    assert math.isclose(mp.mpf(num) / 2 ** mp.mpf(denom), D(num, denom).mpfloat())
    assert math.isclose(num / 2**denom, D(num, denom).mpfloat())


@pytest.mark.parametrize(
    "n1", [D(1, 1), D(0, 0), D(-1, 1), D(-55, 14), D(-759, 30), D(57, 71), 0, 5, -20]
)
@pytest.mark.parametrize("n2", [D(1, 1), D(0, 0), D(-1, 1), D(-55, 14), D(1, 2), D(57, 71)])
def test_summation(n1, n2):
    """Test the addition of two D numbers."""
    n = n1
    n += n2
    assert math.isclose(float(n1 + n2), float(n1) + float(n2)) and math.isclose(
        float(n), float(n1) + float(n2)
    )


@pytest.mark.parametrize(
    "n1", [D(1, 1), D(0, 0), D(-1, 1), D(-55, 14), D(-759, 30), D(1, 2), D(57, 71), 0, 5, -20]
)
@pytest.mark.parametrize("n2", [D(1, 1), D(0, 0), D(-1, 1), D(-55, 14), D(1, 2), D(57, 71)])
def test_subtraction(n1, n2):
    """Test the subtraction of two D numbers."""
    n = n1
    n -= n2
    assert math.isclose(float(n1 - n2), float(n1) - float(n2)) and math.isclose(
        float(n), float(n1) - float(n2)
    )


@pytest.mark.parametrize(
    "n1", [D(1, 1), D(0, 0), D(-1, 1), D(-55, 14), D(-759, 30), D(1, 2), D(57, 71), 0, 5, -20]
)
@pytest.mark.parametrize("n2", [D(1, 1), D(0, 0), D(-1, 1), D(-55, 14), D(1, 2), D(57, 71)])
def test_multiplication(n1, n2):
    """Test the multiplication of two D numbers."""
    n = n1
    n *= n2
    assert math.isclose(float(n1 * n2), float(n1) * float(n2)) and math.isclose(
        float(n), float(n1) * float(n2)
    )


@pytest.mark.parametrize(
    "base", [D(1, 1), D(0, 0), D(-1, 1), D(-55, 14), D(-759, 30), D(1, 2), D(57, 71)]
)
@pytest.mark.parametrize("exponent", [0, 3, 5, 7, 10, 20])
def test_power(base, exponent):
    """Test the power of a D number."""
    n = base
    n **= exponent
    assert math.isclose(float(base**exponent), float(base) ** exponent) and math.isclose(
        float(n), float(base) ** exponent
    )


@pytest.mark.parametrize(
    "n",
    [
        [D(1, 1), "1/2^1"],
        [D(0, 0), "0/2^0"],
        [D(-1, 1), "-1/2^1"],
        [D(-55, 14), "-55/2^14"],
        [D(1, 2), "1/2^2"],
        [D(57, 71), "57/2^71"],
    ],
)
def test_repr(n):
    """Test the string representation of a D number."""
    assert str(n[0]) == n[1]


@pytest.mark.parametrize("not_int", ["1", 1.0, [1], 1.0j])
def test_init_type_error(not_int):
    """Test the raise of a TypeError when the D class is initialized with a non-integer."""
    with pytest.raises(TypeError, match="Class arguments must be of type int, but received"):
        D(not_int, 1)
    with pytest.raises(TypeError, match="Class arguments must be of type int, but received"):
        D(1, not_int)


def test_init_value_error():
    """Test the raise of a ValueError when the D class is initialized with a negative denominator."""
    with pytest.raises(ValueError, match="Denominator exponent must be positive"):
        D(1, -1)


def test_abs():
    """Test the absolute value of a D number."""
    assert abs(D(5, 1)) == D(5, 1) and abs(D(-5, 1)) == D(5, 1)


def test_equality():
    """Test the equality of two D numbers."""
    assert (
        D(1, 1) == D(1, 1)
        and D(1, 1) != D(1, 2)
        and D(1, 1) != D(2, 1)
        and D(0, 0) == 0
        and D(-10, 0) == -10
        and D(1, 1) != 1.0j
    )


def test_inequality():
    """Test the inequality of two D numbers."""
    assert (
        D(1, 1) < D(2, 1)
        and D(1, 2) < D(1, 1)
        and D(1, 1) > D(1, 2)
        and D(2, 1) > D(1, 1)
        and D(1, 1) <= D(1, 1)
        and D(1, 1) >= D(1, 1)
        and D(1, 1) <= D(2, 1)
    )


@pytest.mark.parametrize("nb", [1.0, 1.0j, (1, 1), "1"])
def test_summation_error(nb):
    """Test the raise of a TypeError when the addition of D numbers is not possible."""
    with pytest.raises(TypeError, match="Summation is not defined between D and"):
        D(1, 1) + nb
    with pytest.raises(TypeError, match="Summation is not defined between D and"):
        nb + D(1, 1)


@pytest.mark.parametrize("nb", [1.0, 1.0j, (1, 1), "1"])
def test_subtraction_error(nb):
    """Test the raise of a TypeError when the subtraction of D numbers is not possible."""
    with pytest.raises(TypeError, match="Subtraction is not defined between D and"):
        D(1, 1) - nb


@pytest.mark.parametrize("nb", [1.0, 1.0j, (1, 1), "1"])
def test_multiplication_error(nb):
    """Test the raise of a TypeError when the multiplication of D numbers is not possible."""
    with pytest.raises(TypeError, match="Product is not defined between D and"):
        D(1, 1) * nb
    with pytest.raises(TypeError, match="Product is not defined between D and"):
        nb * D(1, 1)


def test_power_type_error():
    """Test the raise of a TypeError when the power of a D number is not an integer."""
    with pytest.raises(TypeError, match="Expected power to be an integer"):
        D(1, 1) ** 1.0


def test_power_value_error():
    """Test the raise of a ValueError when the power of a D number is negative."""
    with pytest.raises(ValueError, match="Expected power to be a positive integer"):
        D(1, 1) ** -1

File: tests/cpp/test_rings/test_d.cpp
/* Copyright 2024-2025 Olivier Romain, Francis Blais, Vincent Girouard, Marius Trudeau
 * 
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 * 
 *        http://www.apache.org/licenses/LICENSE-2.0
 * 
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */


// Include Libraries
#include <gtest/gtest.h>
#include <boost/multiprecision/cpp_int.hpp>
#include <qdecomp/rings/cpp/Rings.hpp>


// Useful definitions
namespace mp = boost::multiprecision;
using TestingTypes = ::testing::Types<long long int, mp::cpp_int>;

template <typename T>
class DRingTests : public ::testing::Test {};
TYPED_TEST_SUITE(DRingTests, TestingTypes);


// Test addition in D ring
TYPED_TEST(DRingTests, Addition) {
    for (TypeParam a_num : {0, 2, -3}) {
    for (TypeParam b_num : {0, 11, -20}) {
    for (unsigned int a_denom : {0, 1, 2}) {
    for (unsigned int b_denom : {0, 1, 2}) {
        D<TypeParam> a(a_num, a_denom);
        D<TypeParam> b(b_num, b_denom);
        D<TypeParam> result = a + b;
        long double expected = a.to_long_double() + b.to_long_double();
        EXPECT_NEAR(result.to_long_double(), expected, 1e-9);
    }
    }
    }
    }
}

// Test multiplication in D ring
TYPED_TEST(DRingTests, Multiplication) {
    for (TypeParam a_num : {0, 2, -3}) {
    for (TypeParam b_num : {0, 11, -20}) {
    for (unsigned int a_denom : {0, 1, 2}) {
    for (unsigned int b_denom : {0, 1, 2}) {
        D<TypeParam> a(a_num, a_denom);
        D<TypeParam> b(b_num, b_denom);
        D<TypeParam> result = a * b;
        long double expected = a.to_long_double() * b.to_long_double();
        EXPECT_NEAR(result.to_long_double(), expected, 1e-9);
    }
    }
    }
    }
}

// Test equality and inequality in D ring
TYPED_TEST(DRingTests, Equality) {
    for (TypeParam a_num : {0, 2, -3}) {
    for (TypeParam b_num : {0, 11, -20}) {
    for (unsigned int a_denom : {0, 1, 2}) {
    for (unsigned int b_denom : {0, 1, 2}) {
        D<TypeParam> a(a_num, a_denom);
        D<TypeParam> b(b_num, b_denom);

        long double diff = (a - b).to_long_double();
        bool equal = (std::abs(diff) < 1e-9);

        if (equal) {
            EXPECT_TRUE(a == b);
            EXPECT_FALSE(a != b);
        } else {
            EXPECT_FALSE(a == b);
            EXPECT_TRUE(a != b);
        }
    }
    }
    }
    }
}

// Test exponentiation in D ring
TEST(DRingTests, Exponentiation) {
    for (long long int a_num : {0, 2, -3}) {
    for (unsigned int a_denom : {0, 1, 2}) {
    for (unsigned int exp : {0, 1, 2, 3}) {
        D<long long int> a(a_num, a_denom);
        D<long long int> result = a.pow(exp);
        long double expected = std::pow(a.to_long_double(), exp);
        EXPECT_NEAR(result.to_long_double(), expected, 1e-9);
    }
    }
    }
}

// Test the reduce method in D ring
TYPED_TEST(DRingTests, Reduce) {
    for (TypeParam a_num : {0, 2, -3, 4, -16}) {
    for (unsigned int a_denom : {0, 1, 2}) {
        D<TypeParam> a(a_num, a_denom);  // The reduce method is called in the constructor
        bool must_be_reduced = a_num % 2 == 0 && a_denom > 0 && a_num != 0;
        if (must_be_reduced) {EXPECT_NE(a.num(), a_num);}
        else {EXPECT_EQ(a.num(), a_num);}
    }
    }
}
